### PR TITLE
[benchmarks] Benchmark generator & template-heavy benchmark

### PR DESCRIPTION
**Pull request by kevinpschaaf**
Adds a generalized template benchmark generator, and adds a generated `template-heavy` benchmark for lit-html to CI.

For running the generator standalone:
* `npm run generator -- -t` will run generate a default configuration across all supported renderers (lit-html, idom, react, preact) and compare them in tachometer
* `npm run generator -- --help` for full options

For running the generated `template-heavy` benchmark:
* `npm run benchmark:lit-heml:template-heavy` to run locally
* `npm run regen-template-benchmarks` to re-generate the benchmark as checked-in

The `template-heavy` benchmark (as checked in, we can adjust this) was generated with the following parameters:
*  `depth`: 5
*  `width`: 4
*  `attrPerNode`: 4
*  `dynAttrPct`: 0.5 (2 attr per node have bindings, 2 are static)
*  `valPerDynAttr`: 2 (2 parts per attribute binding)
*  `dynNodePct`: 0.5 (half of the nodes in a template are generated with a nested template, half are static)
*  `updateCount`: 10
*  `updateNodePct`: 1 (all nested templates are switched each update)
*  `updateAttrPct`: 1 (all bound attribute values change each update)
*  `uniqueTemplates`: true

Which results in:
* 2072 unique templates (due to dynNodePct, not all templates are used each render)
* 1365 elements / 2907 nodes

---

Update:
* Refactored the deoptigate code and added little CLI for running deoptigate standalone on a given URL under the `/benchmarks` folder using puppeteer with the right flags and log generation:

  ```
  benchmarks$ node generator/build/deoptigate-cli.js -u lit-html/kitchen-sink/index.html -O
  ```

### CODE REVIEW

File: packages/benchmarks/generator/src/index.ts [hunk 1]
```diff
@@ -0,0 +1,327 @@
...
+        s = renderer.openTagStart(s, level, tag, opts.attrPerNode, staticAttrPerNode);
+        for (let j=0; j<opts.attrPerNode; j++) {
+          const isStatic = j < staticAttrPerNode;
+          const isConstant = isStatic || ((j - staticAttrPerNode) < constantAttrPerNode)
+          const name = nextLevel(level, j);
+          s = renderer.setAttr(s, level, name, isStatic, isConstant, Math.max(opts.valPerDynAttr, 1));
+        }
+        s = renderer.openTagEnd(s, level, tag, opts.attrPerNode);
+        s = renderer.textNode(s, level, false, level);
+        if (depthForLevel(level) < opts.depth) {
```
Comment (sorvell) [inline packages/benchmarks/generator/src/index.ts:78]: Hard to follow the code past here. Some comments would be helpful.

**Review by sorvell — CHANGES_REQUESTED** [review]

File: packages/benchmarks/generator/src/deoptigate.ts [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+      `--enable-precise-memory-info`,
+      `--no-sandbox`,
+      `--disable-extensions`,
+      `--js-flags=--trace-ic,--nologfile-per-isolate,--logfile=${logFile}`,
+    ]
+  });
+  console.log(`Profiling ${url}...`);
+  const page = await browser.newPage();
+  await page.goto(`http://localhost:9999/${url}`);
+  await new Promise(r => setTimeout(() => r(), 2000));
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/benchmarks/generator/src/deoptigate.ts:62]: Eh, I just didn't want to implement a polling thing to know when the benchmark was done or the page had settled.

File: packages/benchmarks/generator/src/renderers.ts [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+    },
+    render() {
+      return `patch(container, () => render());`;
+    },
+    endBenchmark() {
+      return `</script>`;
+    },
+  };
+
+  const idomKeyedStaticsRenderer: TemplateRenderer = {
```
[the remark this answers is not in the document]
Comment (kevinpschaaf) [reply packages/benchmarks/generator/src/renderers.ts:154]: In incremental DOM, you can differentiate "static" attributes from dynamic ones and just specify them once, and then you don't have to call `attr` on every render for them, but that also requires keying the nodes. I just did that mode as a separate renderer so we could evaluate the benefit of that optimziation.

http://google.github.io/incremental-dom/#statics-array

**Review by sorvell — APPROVED** [review]
